### PR TITLE
[improve]  Unaware-bucket table do compaction async, uncoupled with CheckpointBarrier.

### PR DESCRIPTION
**Pull request by leaves12138**
Unaware-bucket table do compaction async, uncoupled with CheckpointBarrier.

Cause compaction may stock the checkpoint acknowledgement, but we actually don't need wait for all the compaction tasks finished to complete the checkpoint.

Use the ExecutorService with max 100 tasks queue running background to execute compaction tasks.

<pre>
Worker Task    <---------------------------------------    ResultQueue
     |                                                       /\
     |                       (  o   o   )                    |
     |                            -                          |
     |                                                       |
     |                                                       |
    \/                                                       |
Task Queue  --------------------------------------------> ExecutorService
</pre>

### CODE REVIEW

**Review by JingsongLi — CHANGES_REQUESTED** [review]
Add tests for `AppendOnlyTableCompactionWorkerOperator`.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/AppendOnlyTableCompactionWorkerOperator.java [hunk 1]
```diff
@@ -54,14 +65,27 @@ public AppendOnlyTableCompactionWorkerOperator(
     public void open() throws Exception {
         this.worker = new AppendOnlyTableCompactionWorker(table, commitUser);
         this.result = new ArrayList<>();
+        executor =
+                new ThreadPoolExecutor(
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/AppendOnlyTableCompactionWorkerOperator.java:69]: See `AbstractFileStoreWrite`, how to new an executor and close executor.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/AppendOnlyTableCompactionWorkerOperator.java [hunk 2]
```diff
@@ -54,14 +65,27 @@ public AppendOnlyTableCompactionWorkerOperator(
...
+                        1,
+                        1,
+                        0L,
+                        TimeUnit.MILLISECONDS,
+                        new LinkedBlockingQueue<>(
+                                100)); // most 100 tasks in queue to avoid memory collapse
+        lock = new Object();
     }
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/AppendOnlyTableCompactionWorkerOperator.java:78]: Uncommit files should be deleted in close.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/AppendOnlyTableCompactionWorkerOperator.java [hunk 3]
```diff
@@ -38,10 +44,15 @@
  */
 public class AppendOnlyTableCompactionWorkerOperator
         extends PrepareCommitOperator<AppendOnlyCompactionTask, Committable> {
+    private static final Logger LOG =
+            LoggerFactory.getLogger(AppendOnlyTableCompactionWorkerOperator.class);
     private final AppendOnlyFileStoreTable table;
-    private transient AppendOnlyTableCompactionWorker worker;
     private final String commitUser;
+    private transient AppendOnlyTableCompactionWorker worker;
     private transient List<CommitMessage> result;
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/AppendOnlyTableCompactionWorkerOperator.java:52]: You can just change this to `LinkedBlockingQueue` and avoid lock.

File: paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/FileStoreTableTestBase.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+import org.junit.jupiter.api.io.TempDir;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.UUID;
+
+/** Test base for create table more conveniently. */
+public abstract class FileStoreTableTestBase {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/FileStoreTableTestBase.java:44]: Use `TableTestBase`?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/AppendOnlyTableCompactionWorkerOperator.java [hunk 5]
```diff
@@ -52,25 +70,105 @@ public AppendOnlyTableCompactionWorkerOperator(
...
-        ArrayList<CommitMessage> tempList = new ArrayList<>(result);
-        result.clear();
-        return tempList.stream()
-                .map(s -> new Committable(checkpointId, Committable.Kind.FILE, s))
-                .collect(Collectors.toList());
+        List<CommitMessage> tempList = new ArrayList<>();
+        try {
+            while (!result.isEmpty()) {
+                Future<List<CommitMessage>> future = result.peek();
+                if (!future.isDone()) {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/AppendOnlyTableCompactionWorkerOperator.java:85]:
```
if (!future.isDone() && !waitCompaction) {
    break;
}
....
```

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/AppendOnlyTableCompactionWorkerOperator.java [hunk 6]
```diff
@@ -52,25 +70,105 @@ public AppendOnlyTableCompactionWorkerOperator(
...
+                List<CommitMessage> messages;
+                try {
+                    messages = resultFuture.get();
+                } catch (Exception exception) {
+                    // exception should already be handled
+                    continue;
+                }
+                for (CommitMessage uncommitMessage : messages) {
+                    CommitMessageImpl uncommitMessageImpl = (CommitMessageImpl) uncommitMessage;
+                    List<DataFileMeta> fileMetas =
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/AppendOnlyTableCompactionWorkerOperator.java:159]: See `TableCommit.abort`?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/AppendOnlyTableCompactionWorkerOperator.java (hunk 5 above)
Comment (leaves12138) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/AppendOnlyTableCompactionWorkerOperator.java:85]: reasonable

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**